README.md
# Freeport Smart Contracts

## Introduction

A set of contracts to hold NFTs, capture royalties, and subscribe to services.

See [Freeport.sol](contracts/Freeport.sol).

See the [documentation](docs/Freeport.md) (regenerate with `npm run doc`).

## Deployments


#### v2.0.0 on Polygon (Production)

Contracts: [Freeport](https://polygonscan.com/address/0xf9AC6022F786f6f64Fd8abf661190b8517D92396) and
[Fiat Gateway](https://polygonscan.com/address/0x4478e3B0B71531DAc9d0ECe9357eBB0043669804)
and [Simple Auction](https://polygonscan.com/address/0xf6a530242B233B1b4208c449D6C72FB7c6133cC0)
and [NFT Attachment](https://polygonscan.com/address/0x651f2C6942F1c290632Ad5bB61D9ece789f82f35)
and [USDC ERC20](https://polygonscan.com/address/0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174).

Accounts:
[Fiat-to-NFT Service](https://polygonscan.com/address/0x2352f1167F1c1c5273bB64a1FEEAf9dF49702A19) (
see [prod_setup](scripts/prod_setup.js)). Biconomy enabled.
[Admin](https://polygonscan.com/address/0x87A0892F98914c567379475a806A1419143406F6).

#### v2.0.0 on Polygon Mumbai testnet (Staging)

Contracts: [Freeport](https://mumbai.polygonscan.com/address/0xacd8105cBA046307d2228794ba2F81aA15e82E0D) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0xe25b31Aa454E28110E28C694ded69cD241A27Db1)
and [Simple Auction](https://mumbai.polygonscan.com/address/0x71091eA5466E3Ca50971bd51c11069e5629a49dB)
and [NFT Attachment](https://mumbai.polygonscan.com/address/0x84766787c6b9131927A76634F7DDCfcf3ff2e9d1)
and [TestERC20](https://mumbai.polygonscan.com/address/0x93E73E25F290f8A50281A801109f75CB4E8e3233).

Accounts:
[Fiat-to-NFT Service](https://mumbai.polygonscan.com/address/0x53B53189e668dC2ee3bA7A44Bb033E60F400d395) (
see [stage_setup](scripts/stage_setup.js)). Biconomy enabled.
[Admin](https://mumbai.polygonscan.com/address/0x63846e2D234e4F854F43423014430b4e131f697b).

#### v2.2.0 on Polygon Mumbai testnet (Dev)

Contracts: [Freeport](https://mumbai.polygonscan.com/address/0x702BA959B5542B2Bf88a1C5924F73Ed97482c64B) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0x106Bf3D61952faE9279B08bdcB2e548316E0C1Ae)
and [Simple Auction](https://mumbai.polygonscan.com/address/0xCEAc6c102bEcE4ed2E5ede9df096F7175BB8CbaD)
and [NFT Attachment](https://mumbai.polygonscan.com/address/0x39B27a0bc81C1366E2b05E02642Ef343a4f9223a)
and [TestERC20](https://mumbai.polygonscan.com/address/0x4e5a86E128f8Fb652169f6652e2Cd17aAe409e96).

Accounts:
[Fiat-to-NFT Service](https://mumbai.polygonscan.com/address/0xD2B94CBF0fFAA9bc07126ab53f980Cd95a5Ed243) (
see [dev_setup](scripts/dev_setup.js)). Biconomy enabled.
[Admin](https://mumbai.polygonscan.com/address/0x63846e2D234e4F854F43423014430b4e131f697b).

#### v2.1.0 on Polygon Mumbai (Experiment @AlekseiMisyukevich)

Contracts: [Freeport](https://mumbai.polygonscan.com/address/0xD6e5Ce841f2C044b552DF862Dfb1D936070FC5c6) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0x6546CB4776cF16f565c7D0E38c7ef85534eAE689)
and [Simple Auction](https://mumbai.polygonscan.com/address/0xC4820D5983122618Ef4aBEA9Dc429111e7802003)
and [NFT Attachment](https://mumbai.polygonscan.com/address/0x8001f8229E67Ed35A04fE2A6cA92D28feCD2D85b)
and [TestERC20](https://mumbai.polygonscan.com/address/0x2C390FAB1c3568F783aDEb31F88E1c207ca1E721).

Accounts:
[Fiat-to-NFT Service](https://mumbai.polygonscan.com/address/0xD2B94CBF0fFAA9bc07126ab53f980Cd95a5Ed243) (
see [dev_setup](scripts/dev_setup.js)). Biconomy enabled.
[Admin](https://mumbai.polygonscan.com/address/0x63846e2D234e4F854F43423014430b4e131f697b).

## How to upgrade

Review all code changes since the last deployed release. Understand how upgradeable contracts work (see [doc](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable)). Pay attention to:
- Events: make sure the event listeners are upgraded to support **both new and old** versions of events at the same time.
- Structure changes: make sure existing fields do not change in any way, only new fields are added.
- Existing data: the existing objects in the contract will have a zero value for new fields; make sure the code handles this correctly.
- Functions: make sure that no public function was removed or signature changed.
- Initialization: if a new feature requires some initialization, make sure this happens in a new `initialize_vX.X.X` function, and that this function is called during the upgrade (example: [initialize_v2_0_0](https://github.com/Cerebellum-Network/Freeport-Smart-Contracts/blob/bc0aae14cc6d0dbbdb3656b6aba6e15085b03fbf/migrations/9_upgrade_auction.js#L13)).
- Unit tests: understand what behavior changed.

Bump the version, maintain [CHANGELOG.md](CHANGELOG.md), and create a release in github.

Choose a [migration script](migrations/) depending on which contract to upgrade (e.g.: N=8). For the DEV environment, you need to edit the addresses in `build/*.json`. Then run:

    npm run test
    export MNEMONIC="  …mnemonic of a wallet that is admin…  "
    truffle migrate --compile-all --network=polygon_testnet -f N --to N

Release the JS SDK (see [instructions](https://github.com/Cerebellum-Network/Freeport-Smart-Contracts-SDK#publishing-to-npm)).

## How to deploy

    npm test
    npm run migrate --network=polygon_testnet

Write down the contract address and current version in the Releases section above.

Commit these files in `build/contracts/`: `Freeport.json`, `FiatGateway.json`, `SimpleAuction.json`
, `NFTAttachment.json`.

For **dev** on Polygon testnet, setup test accounts:

    truffle exec scripts/dev_setup.js --network=polygon_testnet

For **staging** on Polygon testnet, setup test accounts:

    truffle exec scripts/stage_setup.js --network=polygon_testnet

For **prod** on Polygon mainnet, setup production accounts (edit the scripts to change accounts):

    truffle exec scripts/prod_setup.js --network=polygon_mainnet
    truffle exec scripts/prod_add_admin.js --network polygon_mainnet

The script will print the address of the different accounts, including the fiat-to-nft service account.

Verify on Polyscan using the information in Freeport.json and the flattened code:

    npm run flatten

More instructions can be found in the original [README of the template](BUILD.md).

## How to use test image

### 1. To copy the latest snapshot of db with deployed contract and built contracts use

`sudo . ./copy-artifacts.sh dir_path_to (copy to your directory)`\
`sudo . ./copy-artifacts.sh (copy to default ./artifacts directory)`

### 2. To run ganache cli locally with already deployed contract use

`export MNEMONIC='spatial spin account funny glue cancel cushion twelve inmate author night dust'`\
`export NETWORK_ID=5777`\
`docker run -d -p 8545:8545 524725240689.dkr.ecr.us-west-2.amazonaws.com/crb-davinci-nft-test:latest --db /app/db --mnemonic $MNEMONIC --networkId $NETWORK_ID`

## How to update Freeport SDK

1. Open [SDK](https://github.com/Cerebellum-Network/Freeport-Smart-Contracts-SDK) project in IDE, switch to master
   branch
2. Copy smart-contract artifacts from `/build/contracts` to SDK's `/src/artifacts`
3. Run `yarn update-types`
4. Commit changes with message like *"artifacts from commit e57691bc"*
5. Check that `npm whoami` returns "cere-io" (otherwise update your NPM_TOKEN env var for publishing @cere packages)
6. Run `yarn deploy` and publish next version (
   see [Versioning](https://github.com/Cerebellum-Network/Freeport-Smart-Contracts-SDK#versioning))
7. Push changes to the remote branch
8. Install the new version of SDK in projects that use it

## How to run e2e test locally

Readme.md for running e2e-test locally
is [here](https://github.com/Cerebellum-Network/e2e-tests/blob/master/README.md#how-to-run-e2e-tests-locally). For this
service you need to create image locally and tag it with
tag `524725240689.dkr.ecr.us-west-2.amazonaws.com/crb-davinci-nft-test:YOUR_CUSTOM_TAG`
and use `YOUR_CUSTOM_TAG` for running e2e-tests locally.

## Where to check the e2e-tests result after merge in dev?

After merge in develop, make sure that tests passed in **@e2e-test-results** channel on Cere Slack.


CHANGELOG.md
### Next release

- Sale SC created based on SimpleExchange. Its purpose is to run nft sale at fix price.
- Added ERC20 to Sale instead of internal currency for payments.
- Fees calculation implemented in Freeport.
- Fees capture is delegated to Sale.
- Added quantity param to makeOffer.

### v2.2.0

- Support for private auctions. Bidders of a private auction must obtain an authorization from a central service.
- Distribute the shares of joint accounts directly in USDC.

### v2.0.0

- Make SimpleAuction, SimpleExchange, and FiatGateway work with ERC20 directly.
- Remove the function FiatGateway.buyCereFromUsd.

### v1.1.0

- Fix a bug of incorrect royalty collection in auctions.
- Fix a vulnerability that may allow a buyer to disrupt an auction.
- Additional public function transferFrom has been implemented in Freeport contract to support the above fixes.

### 2022-01-28: Upgradeability

Make Freeport, FiatGateway, SimpleAuction, and NFTAttachment upgradeable.

### 2022-01-07: Extended NFT Attachments

NFT Attachments now support variable-sized attachment data, and separate events for minter, owner, and anonymous senders. This is included in the deployment 2021-12-31 below.

### 2021-12-31: ERC20 Adapter

Use an ERC20 adapter instead of a Polygon bridge adapter.
Commit caefd460 deployed.

#### Mainnet

Contracts: [Freeport](https://polygonscan.com/address/0x521296be238B164b9A391b6F6175741826CB5F33) and
[Fiat Gateway](https://polygonscan.com/address/0xf6d782Cd0dC9976170242B94C8E653C7bA489634) and [Simple Auction](https://polygonscan.com/address/0x166996399a262d8F0976c2fa206E92F8BE77810F) and [NFT Attachment](https://polygonscan.com/address/0x5E376313fddBE3010F5d9fbC446C63d803590445) and [Forwarder](https://polygonscan.com/address/0x5e43bA1666B13C346E2ECdeD9dcDaaf02fbb0B22) and [Bypass Forwarder](https://polygonscan.com/address/0x1E77956B211cb4437317CF692141b292B1433f29) and [USDC ERC20](https://polygonscan.com/address/0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174).

Accounts:
[Fiat-to-NFT Service](https://polygonscan.com/address/0x2352f1167F1c1c5273bB64a1FEEAf9dF49702A19) (
see [prod_setup](scripts/prod_setup.js)).
No Meta-tx Relayer Account yet.
[Admin](https://polygonscan.com/address/0x4d632f8513554d7647a326c1f4ca72caeff93e63).

#### Stage

Contracts: [Freeport](https://mumbai.polygonscan.com/address/0x8bD1D3a93C7FB1786fFE3d0610987C3879287698) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0x1C59A68ff017f14D1a8B80644F25F047b1CC58C5) and [Simple Auction](https://mumbai.polygonscan.com/address/0xCEfC0a243D56fa9A986D173f6289EcBBD336643B) and [NFT Attachment](https://mumbai.polygonscan.com/address/0x64fe48A0555b3b822E5DAC8347DFD1FDc9A2E91D) and [Forwarder](https://mumbai.polygonscan.com/address/0x03988B5eaf8EFC804320B860dBb7f281EdF92420) and [Bypass Forwarder](https://mumbai.polygonscan.com/address/0x7a770bf5a93a3a686FA2B40cA399462ceD10725D) and [TestERC20](https://mumbai.polygonscan.com/address/0x93E73E25F290f8A50281A801109f75CB4E8e3233).

Accounts:
[Fiat-to-NFT Service](https://mumbai.polygonscan.com/address/0x53B53189e668dC2ee3bA7A44Bb033E60F400d395) (
see [stage_setup](scripts/stage_setup.js)).
No Meta-tx Relayer Account yet.
[Admin](https://mumbai.polygonscan.com/address/0x63846e2D234e4F854F43423014430b4e131f697b).


#### Dev

Contracts: [Freeport](https://mumbai.polygonscan.com/address/0xC59Af7FbE4553e07aA668C1A13CAa78Cd4550579) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0xE8949692827C3034c6fF185a38c192ca3059f6e5) and [Simple Auction](https://mumbai.polygonscan.com/address/0x9E3c717d9fD5839131e567fCAf755c979bF78D08) and [NFT Attachment](https://mumbai.polygonscan.com/address/0x5d0a9933D779265B67429147184261eB7163370b) and [Forwarder](https://mumbai.polygonscan.com/address/0x2d7FCbBfe773c5E7C7fccbA8434386048267c16D) and [Bypass Forwarder](https://mumbai.polygonscan.com/address/0xF7544C67e382230B2732C8360BfAEeAE840C8b1e) and [TestERC20](https://mumbai.polygonscan.com/address/0x4e5a86E128f8Fb652169f6652e2Cd17aAe409e96).

Accounts:
[Fiat-to-NFT Service](https://mumbai.polygonscan.com/address/0xD2B94CBF0fFAA9bc07126ab53f980Cd95a5Ed243) (
see [dev_setup](scripts/dev_setup.js)).
No Meta-tx Relayer Account yet.
[Admin](https://mumbai.polygonscan.com/address/0x63846e2D234e4F854F43423014430b4e131f697b).


### 2021-11-08: CID Attachments

- Attach arbitrary objects to NFTs.

NFTAttachment commit 89696561 deployed for Freeport version 2021-09-13 and 2021-11-03 below.

### 2021-11-04: Auction with getter

- Add the getter for bids (sellerNftBids).
- Prevent potential reentrancy bug.

SimpleAuction commit a74e4db0 deployed for Freeport version 2021-09-13 and 2021-11-03 below.

### 2021-11-03: Auction with royalties (dev deployment)

- Auction sales now capture royalties, if any.

Commit 6df6c349 deployed on Polygon Mumbai.

Contracts [Freeport](https://mumbai.polygonscan.com/address/0xd1EdBAC660307c5B6d22E678FB5e22668C70Ad96) and
[Fiat Gateway](https://mumbai.polygonscan.com/address/0x1f8eC932B6ec39A0326b74E9648A158F88B24082) and [Simple Auction](https://mumbai.polygonscan.com/address/0xd7cd23C84F9109F57f13eF28319e8787628DD7ad) and [NFT Attachment](https://mumbai.polygonscan.com/address/0x270693f873287a39172856Ad8cfbCd79b040b287).

[Fiat-to-NFT Service Account](https://mumbai.polygonscan.com/address/0xD2B94CBF0fFAA9bc07126ab53f980Cd95a5Ed243) (
see [dev_setup](scripts/dev_setup.js)).
[Admin Account](https://mumbai.polygonscan.com/address/0x51c5590504251A5993Ba6A46246f87Fa0eaE5897) (Aurel).

Exchange rate of 0.1 CERE_stage for $0.01.

### 2021-10-21: Simple Auction

- A contract that holds auctions and make transfers in the main Freeport contract.
- An external function `captureFee` on the Freeport contract that can be used by authorized contracts (i.e., the auction contract).

scripts/dev_setup.js
const Freeport = artifacts.require("Freeport");
const TestERC20 = artifacts.require("TestERC20");
const FiatGateway = artifacts.require("FiatGateway");
const SimpleAuction = artifacts.require("SimpleAuction");
const {getSigner} = require("../test/utils");
const log = console.log;

module.exports = async function (done) {

    // A fixed account for tests.
    let serviceAccount = "0x0EA2A95dE281826d5B30587C2bA2218d97476008";

    const CURRENCY = 0;
    let tenM = "10" + "000" + "000" + "000000"; // 10M with 6 decimals;
    let oneM = "1" + "000" + "000" + "000000"; // 1M with 6 decimals;
    
    const signer = await getSigner();
    let authorizer = signer.address;
    let accounts = await web3.eth.getAccounts();
    let admin = accounts[0];
    let freeport = await Freeport.at("0x702BA959B5542B2Bf88a1C5924F73Ed97482c64B");
    let erc20 = await TestERC20.at("0x4e5a86E128f8Fb652169f6652e2Cd17aAe409e96");
    let gateway = await FiatGateway.at("0x106Bf3D61952faE9279B08bdcB2e548316E0C1Ae");
    let auction = await SimpleAuction.at("0x7e4FCB28B5794dBf729E860f0abe97C3412E62e4");
    log("Operating on Freeport contract", freeport.address);
    log("Operating on TestERC20 contract", erc20.address);
    log("Operating on FiatGateway contract", gateway.address);
    log("Operating on SimpleAuction contract", auction.address);
    log("From admin account", admin);
    log("With Authorizer account", authorizer.address);
    
    log("Set an exchange rate of 1:1 (0.01 with 6 decimals of ERC20 for $0.01)");
    await gateway.setExchangeRate(0.01e6);

    log("Give the permission to execute payments to the service account", serviceAccount);
    const PAYMENT_SERVICE = await gateway.PAYMENT_SERVICE.call();
    await gateway.grantRole(PAYMENT_SERVICE, serviceAccount);

    log("Grant role to address for bid authorization");
    const BUY_AUTHORIZER_ROLE = await auction.BUY_AUTHORIZER_ROLE.call();
    await auction.grantRole(BUY_AUTHORIZER_ROLE, authorizer);

    await erc20.mint(gateway.address, tenM);
    log("Sent 10M of currency to FiatGateway");

    let devAccounts = [
        admin,
        "0x6108E8aFFe0c51D4e2515A773aeF16b19ED6feB9", // e2e tests (Pavel)
        "0x6d2b28389d3153689c57909829dFCf6241d36388", // Evgeny
        "0x1Bf6FCa28253A1257e4B5B3440F7fbE0c59D1546", // Sergey
        "0x51c5590504251A5993Ba6A46246f87Fa0eaE5897", // Aurel
    ];

    for (let devAccount of devAccounts) {
        await erc20.mint(devAccount, oneM);
        log("Sent 1M of currency to", devAccount);
    }

    done();
};